src/schedule_systems.rs

```diff
@@ -85,75 +82,90 @@ pub(crate) fn run_synctest<C: Config>(world: &mut World, mut sess: SyncTestSessi
-            Err(GGRSError::PredictionThreshold) => {
-                info!("P2PSpectatorSession: Waiting for input from host.")
-            }
-            Err(e) => warn!("{e}"),
-        };
-    }
+    let running = sess.current_state() == SessionState::Running;
+    let requests = running.then(|| sess.advance_frame());
 
     world.insert_resource(Session::Spectator(sess));
+
+    match requests {
+        Some(Ok(requests)) => handle_requests(requests, world),
+        Some(Err(GGRSError::PredictionThreshold)) => {
+            info!("P2PSpectatorSession: Waiting for input from host.")
+        }
+        Some(Err(e)) => warn!("{e}"),
+        None => {}
+    };
 }
 
 pub(crate) fn run_p2p<C: Config>(world: &mut World, mut sess: P2PSession<C>) {
-    maybe_init_snapshots(world, sess.max_prediction());
-
     world.insert_resource(LocalPlayers(sess.local_player_handles()));
 
-    if sess.current_state() == SessionState::Running {
+    let running = sess.current_state() == SessionState::Running;
+
+    if running {
         // get local player inputs
         world.run_schedule(ReadInputs);
+
         let local_inputs = world.remove_resource::<LocalInputs<C>>().expect(
             "No local player inputs found. Did you insert systems into the ReadInputs schedule?",
         );
+
         for (handle, input) in local_inputs.0 {
             sess.add_local_input(handle, input)
                 .expect("All handles in local_inputs should be valid");
         }
-
-        match sess.advance_frame() {
-            Ok(requests) => handle_requests(requests, world),
-            Err(GGRSError::PredictionThreshold) => {
-                info!("Skipping a frame: PredictionThreshold.")
-            }
-            Err(e) => warn!("{e}"),
-        };
     }
 
+    let requests = running.then(|| sess.advance_frame());
+
     world.insert_resource(Session::P2P(sess));
+
+    match requests {
+        Some(Ok(requests)) => handle_requests(requests, world),
+        Some(Err(GGRSError::PredictionThreshold)) => {
+            info!("Skipping a frame: PredictionThreshold.")
+        }
+        Some(Err(e)) => warn!("{e}"),
+        None => {}
+    }
 }
 
 pub(crate) fn handle_requests<T: Config>(requests: Vec<GGRSRequest<T>>, world: &mut World) {
     for request in requests {
+        if let Some(Session::P2P(s)) = world.get_resource::<Session<T>>() {
+            let confirmed_frame = s.confirmed_frame();
+            world.insert_resource(RollbackFrameConfirmed(confirmed_frame));
+        }
+
         match request {
             GGRSRequest::SaveGameState { cell, frame } => {
                 debug!("saving snapshot for frame {frame}");
                 world.run_schedule(SaveWorld);
 
                 // look into resources and find the checksum
                 let snapshots = world
-                    .get_resource::<RollbackSnapshots>()
+                    .get_resource::<GgrsSnapshots<Checksum, Option<Checksum>>>()
```

Comment:
Better to just get the `Checksum` resource directly here?

Reply:
Totally right! I forgot that the `frame` value provided in `SaveGameState` is supposed to be the current frame anyway.